File: backend/src/main/java/kg/taskflow/service/FileService.java
package kg.taskflow.service;

import org.springframework.web.multipart.MultipartFile;

import java.util.List;

/**
 * Service for file operations with MinIO
 */
public interface FileService {

    /**
     * Upload a file
     * @param file the file to upload
     * @param path optional path prefix (e.g., "tasks/123", "avatars")
     * @return the URL of the uploaded file
     */
    String uploadFile(MultipartFile file, String path);

    /**
     * Upload multiple files
     */
    List<String> uploadFiles(List<MultipartFile> files, String path);

    /**
     * Upload a file from Base64 data
     */
    String uploadBase64File(String base64Data, String fileName, String path);

    /**
     * Download a file as bytes
     */
    byte[] downloadFile(String filePath);

    /**
     * Delete a file
     */
    void deleteFile(String filePath);

    /**
     * Get a presigned URL for temporary access
     * @param filePath the file path
     * @param expirySeconds URL expiry in seconds
     */
    String getPresignedUrl(String filePath, int expirySeconds);

    /**
     * Check if a file exists
     */
    boolean fileExists(String filePath);

    /**
     * List files by prefix
     */
    List<String> listFiles(String prefix);

    /**
     * Get the public URL for a file
     */
    String getPublicUrl(String filePath);
}

File: frontend/src/services/file.service.ts
import api from './api';

export interface FileUploadResponse {
  success: boolean;
  filePath: string;
  url: string;
  originalName?: string;
  size?: number;
  contentType?: string;
}

export interface MultipleFilesUploadResponse {
  success: boolean;
  count: number;
  filePaths: string[];
  urls: string[];
}

export interface PresignedUrlResponse {
  success: boolean;
  url: string;
  expiresIn: number;
}

export const fileService = {
  async uploadFile(file: File, path?: string): Promise<FileUploadResponse> {
    const formData = new FormData();
    formData.append('file', file);
    if (path) {
      formData.append('path', path);
    }

    const response = await api.post<FileUploadResponse>('/files/upload', formData, {
      headers: {
        'Content-Type': 'multipart/form-data',
      },
    });
    return response.data;
  },

  async uploadFiles(files: File[], path?: string): Promise<MultipleFilesUploadResponse> {
    const formData = new FormData();
    files.forEach((file) => {
      formData.append('files', file);
    });
    if (path) {
      formData.append('path', path);
    }

    const response = await api.post<MultipleFilesUploadResponse>('/files/upload/multiple', formData, {
      headers: {
        'Content-Type': 'multipart/form-data',
      },
    });
    return response.data;
  },

  async uploadBase64(data: string, fileName: string, path?: string): Promise<FileUploadResponse> {
    const response = await api.post<FileUploadResponse>('/files/upload/base64', {
      data,
      fileName,
      path,
    });
    return response.data;
  },

  async getPresignedUrl(filePath: string, expirySeconds = 3600): Promise<PresignedUrlResponse> {
    const response = await api.get<PresignedUrlResponse>('/files/presigned-url', {
      params: { path: filePath, expiry: expirySeconds },
    });
    return response.data;
  },

  async deleteFile(filePath: string): Promise<void> {
    await api.delete('/files', { params: { path: filePath } });
  },

  async listFiles(prefix?: string): Promise<string[]> {
    const response = await api.get<{ files: string[] }>('/files/list', {
      params: { prefix },
    });
    return response.data.files;
  },

  getViewUrl(filePath: string): string {
    return `${api.defaults.baseURL}/files/view?path=${encodeURIComponent(filePath)}`;
  },

  getDownloadUrl(filePath: string): string {
    return `${api.defaults.baseURL}/files/download?path=${encodeURIComponent(filePath)}`;
  },
};
